[ci skip] Add documentation for configuring non-primary databases using an environment variable

floehopper opened the pull request:
This documents the behaviour added in #36773 which allows setting a database connection URL in an environment variable named `#{name.upcase}_DATABASE_URL`.

Unless the `--skip-solid` option to `rails new` is set, an explanatory comment is also added to the `database.yml` config file to complement the solid-related multi-database configurations.

skipkayhil commented on railties/lib/rails/generators/rails/app/templates/config/databases/trilogy.yml.tt:
```diff
@@ -51,6 +51,12 @@ test:
 #   production:
 #     url: <%%= ENV["MY_APP_DATABASE_URL"] %>
 #
+# The connection URL for non-primary databases can also be configured using a
+# similar environment variable prefixed with the configuration key in uppercase.
+# For example:
+#
+#   CACHE_DATABASE_URL="trilogy://cacheuser:cachepass@localhost/cachedatabase"
+#
```
I think this is generally a good change, but it seems unnecessary with `--skip-solid`. What do you think about making it conditional on the solid databases being used?

floehopper replied:
Makes sense to me. I'm AFK
at the moment, but I'll update the branch as soon as I have time - probably tomorrow morning.

floehopper replied:
Addressed in https://github.com/rails/rails/pull/54765/commits/8a11d7c55d17e584140951ed9e2af53f0e6d72d6. I've also updated the PR description & the commit note. Marking as resolved.

floehopper commented:
@skipkayhil It's not obvious to me that the build failure is related to the changes in this PR. Is it some kind of intermittent failure? And if so, is there a way to trigger another build?

skipkayhil commented:
> @skipkayhil It's not obvious to me that the build failure is related to the changes in this PR. Is it some kind of intermittent failure? And if so, is there a way to trigger another build?

Yeah I agree it was a flake, so I hit rebuild 👍

floehopper commented:
@skipkayhil  The `docs-preview` build has failed again for the same reason:

```
Unable to find image '973266071021.dkr.ecr.us-east-1.amazonaws.com/builds:ruby-3-3-br-main' locally
docker: Error response from daemon: manifest for 973266071021.dkr.ecr.us-east-1.amazonaws.com/builds:ruby-3-3-br-main not found: manifest unknown: Requested image not found.
```

However, it looks as if all of the `docs-preview` builds run today have failed with the same error. Do you have any idea what's wrong?

zzak commented:
> Unable to find image '973266071021.dkr.ecr.us-east-1.amazonaws.com/builds:ruby-3-3-br-main' locally
This happens when we haven't had a build on `main` in a while... I think it will be fixed in rails/buildkite-config#142 where I just added a builder step at the beginning of the job.

floehopper committed:
Document configuring non-primary db with env var

This documents the behaviour added in #36773 which allows setting a
database connection URL in an environment variable named
"#{name.upcase}_DATABASE_URL".

Unless the `--skip-solid` option to `rails new` is set, an explanatory
comment is also added to the `database.yml` config file to complement
the solid-related database configurations.

floehopper commented:
> Edit: hmm still flaking... I can try hitting rebuild in a few other places or you could also try force pushing the branch which would do the same thing 🤷 But also, this is a small doc change so I don't think `docs-preview` passing is strictly necessary

I just tried force-pushing, but the same build has failed again in the same way. Let me know if there's anything else I can do to progress this.

zzak commented:
Ok: https://freerange-add-docs-for-confi.rails-docs-preview.pages.dev

For now we have to manually trigger a build on main that pushes the Ruby 3.3 image:
https://buildkite.com/rails/rails/builds/117571#0195a892-3e87-4937-9881-c0eb77a47421

Hopefully we can fix this soon :pray:

floehopper commented:
@zzak Thank you! ❤️

eileencodes reviewed (changes requested):
As the author of this feature, I'm a little particular about how we talk about multi-db connections so I've requested changes. Thanks for working on this!

floehopper commented:
> As the author of this feature, I'm a little particular about how we talk about multi-db connections so I've requested changes. Thanks for working on this!

That's great - I really appreciate your feedback! 😄

floehopper commented on guides/source/active_record_multiple_databases.md:
```diff
@@ -86,6 +86,13 @@ production:
+it's possible to configure other databases using an environment variable prefixed with the
+configuration key in uppercase. For example, setting:
+`ANIMALS_DATABASE_URL="mysql2://username:password@host/database"` overrides the `animals`
+configuration in the "production" environment.
```
> I feel like this documentation is confusing to read because it's a bit of a run on sentence.

Reading it back, I agree that this could be improved.

I noticed that you didn't highlight these two lines in your comment above:

> NOTE: In the same way that it's possible to [configure](configuring.html#configuring-a-database)
the primary database by setting the connection string in the `DATABASE_URL` environment variable,

Was that intentional? Do you think those lines are OK or do you think they should not be included?

> Also the `DATABASE_URL` env var doesn't override, it merges so I changed that to be more accurate.

This is actually why I had included the link to the [Configuring a database](https://guides.rubyonrails.org/configuring.html#configuring-a-database) section of the Rails guide which explains in more detail how the env var is combined with the `database.yml` config. However, I think you're right that it would be good to be very explicit that the env var merges.

floehopper replied:
I've attempted to address your concerns about this in https://github.com/rails/rails/pull/54765/commits/dcbf3b92fca9fe1f229f62f64595137fe03c923c. Let me know what you think.

floehopper commented on railties/lib/rails/generators/rails/app/templates/config/databases/mysql.yml.tt:
```diff
@@ -49,6 +49,14 @@ test:
 #
+<%- unless options.skip_solid? -%>
+# The connection URL for non-primary databases can also be configured using a
+# similar environment variable prefixed with the configuration key in uppercase.
```
> It's more correct to refer to that key as the connection name, because it's 3-tier and the key is development and animals. I'm not trying to be pedantic but I'm particular about this because I don't want multiple names for the same thing in our docs.

That makes complete sense. However, I also find the revised version slightly confusing because it talks about both prefixing and appending. How about changing the original to this:

```
# Connection URLs for non-primary databases can also be configured using
# environment variables. The variable name is formed by concatenating the
# connection name with `_DATABASE_URL`.
# For example:
#
#   CACHE_DATABASE_URL="mysql2://cacheuser:cachepass@localhost/cachedatabase"
#
```

floehopper replied:
I've attempted to address your concerns in https://github.com/rails/rails/pull/54765/commits/689188a5a9d9e3f7d6b60a3958b84be865384335. Let me know what you think.

floehopper commented on railties/lib/rails/generators/rails/app/templates/config/databases/postgresql.yml.tt:
```diff
@@ -81,6 +81,14 @@ test:
 #
+<%- unless options.skip_solid? -%>
+# The connection URL for non-primary databases can also be configured using a
+# similar environment variable prefixed with the configuration key in uppercase.
```
@eileencodes I've attempted to address your concerns in https://github.com/rails/rails/commit/689188a5a9d9e3f7d6b60a3958b84be865384335. Let me know what you think.

floehopper commented on railties/lib/rails/generators/rails/app/templates/config/databases/trilogy.yml.tt:
```diff
@@ -51,6 +51,14 @@ test:
 #
+<%- unless options.skip_solid? -%>
+# The connection URL for non-primary databases can also be configured using a
+# similar environment variable prefixed with the configuration key in uppercase.
```
@eileencodes I've attempted to address your concerns in https://github.com/rails/rails/commit/689188a5a9d9e3f7d6b60a3958b84be865384335. Let me know what you think.

floehopper commented:
@eileencodes I've attempted to address your concerns in a couple of fixup commits which I've now combined into a single commit and force-pushed. I'd really appreciate your feedback and hopefully get these changes merged. Thanks!

floehopper commented:
@skipkayhil Is there anything else I can do to move this forward?

floehopper commented:
Any progress on this? Thanks.

floehopper commented:
Thank you!
